### PR TITLE
[NCL-1984] Improve exception handling

### PR DESCRIPTION
**Pull request by stanlyDoge**

### CODE REVIEW

**Comment by janinko** [comment]
Add to whitelist

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 1]
```diff
@@ -249,8 +257,10 @@ public Response removeProduct(
             response.setSuccess(productService.removeProduct(product.getName(),
                     product.getVersion()));
         } catch (EntityNotFoundException e) {
-            return Response.status(Response.Status.NOT_FOUND)
-                    .entity(new ErrorMessage("Product not found")).build();
+            return Response
+                    .status(Response.Status.NOT_FOUND)
+                    .entity(new ErrorMessage(ErrorMessage.eType.PRODUCT_NOT_FOUND,
+                            "Product not found", e.getMessage())).build();
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: Why not use ``PRODUCT_NOT_FOUND`` here?

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 2]
```diff
@@ -168,16 +172,18 @@ public Response addWhiteArtifact(
...
                     response.setSuccess(false);
                     return Response.ok(response).build();
                 default:
                     response.setSuccess(false);
-                    return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
-                            .entity(new ErrorMessage("Unexpected server error occurred.")).build();
+                    return Response
+                            .status(Response.Status.INTERNAL_SERVER_ERROR)
+                            .entity(new ErrorMessage(ErrorMessage.eType.UNEXPECTED_SERVER_ERR,
+                                    "Unexpected server error occurred.")).build();
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: The ``response`` is not used here, so the ``response.setSuccess(false);`` is unnecessary.

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 3]
```diff
@@ -140,8 +142,10 @@ public Response fillFromGAVBom(
                 response.setMessage("Error while analysing pom file");
                 return Response.status(Status.BAD_REQUEST).entity(response).build();
         }
-        return Response.status(Status.INTERNAL_SERVER_ERROR)
-                .entity(new ErrorMessage("There was some internal error")).build();
+        return Response
+                .status(Status.INTERNAL_SERVER_ERROR)
+                .entity(new ErrorMessage(ErrorMessage.eType.INTERNAL_ERR,
+                        "There was some internal error")).build();
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: Make this a default case in the switch and return unexpected server error. See ``addWhiteArtifact`` below.

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 4]
```diff
@@ -112,8 +112,10 @@ public Response fillFromGitBom(
...
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: Make this a default case in the switch and return unexpected server error. See ``addWhiteArtifact`` below.

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 5]
```diff
@@ -432,8 +448,10 @@ public Response addBlackArtifact(
                 return Response.ok(response).build();
             default:
                 response.setSuccess(false);
-                return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
-                        .entity(new ErrorMessage("Unexpected server error occurred.")).build();
+                return Response
+                        .status(Response.Status.INTERNAL_SERVER_ERROR)
+                        .entity(new ErrorMessage(ErrorMessage.eType.UNEXPECTED_SERVER_ERR,
+                                "Unexpected server error occurred.")).build();
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: the ``response.setSuccess(false);`` is unnecessary here.

File: testsuite/src/test/java/org/jboss/da/test/client/rest/bc/RestApiStartProcessRemoteTest.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ *
+ * @author sknot
+ */
+public class RestApiStartProcessRemoteTest extends AbstractRestBCTest {
+
+    @Test
+    public void testStartProcess() throws Exception {
+        ClientResponse<String> response = assertResponseForRequest(RestApiBCTest.PATH_BC_START,
+                "start-process");
```
Comment (janinko) [inline testsuite/src/test/java/org/jboss/da/test/client/rest/bc/RestApiStartProcessRemoteTest.java]: Can you add the assert for response code?

File: testsuite/src/test/java/org/jboss/da/test/client/rest/reports/AlignRemoteTest.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ *
+ * @author sknot
+ */
+public class AlignRemoteTest extends AbstractRestReportsTest {
+
+    @Test
+    public void test() throws Exception {
+        ClientResponse<String> response = assertResponseForRequest(
+                RestApiReportsRemoteTest.PATH_REPORTS_ALIGN, "align-test");
```
Comment (janinko) [inline testsuite/src/test/java/org/jboss/da/test/client/rest/reports/AlignRemoteTest.java]: Can you add the assert for response code?

**Comment by janinko** [comment]
Also please change these types:
COMM_FAIL - no abbreviation
NAME_AND_VERSION_REQUIRED, ALL_PARAMS_REQUIRED - unify the types, keep the messages
ILLEGAL_ARGS - no abbreviation
COMM_REPO_FAIL - merge into COMM_FAIL

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 8]
```diff
@@ -111,9 +111,12 @@ public Response fillFromGitBom(
                 response.setSuccess(false);
                 response.setMessage("Error while analysing pom file");
                 return Response.status(Status.BAD_REQUEST).entity(response).build();
+            default:
+                return Response
+                        .status(Status.INTERNAL_SERVER_ERROR)
+                        .entity(new ErrorMessage(ErrorMessage.eType.INTERNAL_ERR,
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: Thsi should be also UNEXPECTED_SERVER_ERR

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java [hunk 9]
```diff
@@ -165,19 +170,18 @@ public Response addWhiteArtifact(
...
                 case IS_BLACKLISTED:
-                    response.setSuccess(false);
                     return Response
                             .status(Response.Status.CONFLICT)
-                            .entity(new ErrorMessage(
+                            .entity(new ErrorMessage(ErrorMessage.eType.BLACKLIST,
                                     "Can't add artifact to whitelist, artifact is blacklisted"))
                             .build();
                 case NOT_MODIFIED:
-                    response.setSuccess(false);
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: This should stay here.

File: reports-model/src/main/java/org/jboss/da/model/rest/ErrorMessage.java [hunk 10]
```diff
@@ -11,7 +11,30 @@
...
-    private final String message;
+    private final String errorMessage;
+
+    @Getter
+    private final Object details;
+
+    /*
+     * public ErrorMessage(String message, eType errType, Object details) { this.message = message; this.errType = errType;
+     * this.details = details; }
+     */
```
Comment (janinko) [inline reports-model/src/main/java/org/jboss/da/model/rest/ErrorMessage.java]: Why is there a commented code?

File: reports-model/src/main/java/org/jboss/da/model/rest/ErrorMessage.java [hunk 11]
```diff
@@ -11,7 +11,30 @@
 @RequiredArgsConstructor
 public class ErrorMessage {
 
+    public static enum eType {
+        INTERNAL_ERR, BLACKLIST, UNEXPECTED_SERVER_ERR, PRODUCT_NOT_FOUND, ALL_PARAMS_REQUIRED,
```
Comment (janinko) [inline reports-model/src/main/java/org/jboss/da/model/rest/ErrorMessage.java]: ``INTERNAL_ERR`` is not used anymore

File: reports-rest/src/main/java/org/jboss/da/rest/reports/Reports.java [hunk 12]
```diff
@@ -240,9 +278,22 @@ public Response builtReport(BuiltReportRequest request) {
...
             Set<BuiltReportModule> builtReport = reportsGenerator.getBuiltReport(locator);
             return Response.status(Status.OK).entity(toBuiltReport(builtReport)).build();
-        } catch (ScmException | PomAnalysisException | CommunicationException e) {
+        } catch (ScmException e) {
+            log.error("Exception thrown in scm endpoint", e);
+            return Response.status(Status.INTERNAL_SERVER_ERROR)
+                    .entity(new ErrorMessage(ErrorMessage.eType.SCM_ENDPOINT, e.getMessage()))
+                    .build();
+        } catch (PomAnalysisException e) {
             log.error("Exception thrown in scm endpoint", e);
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/reports/Reports.java]: This is exception in POM Analysis, not SCM

File: reports-rest/src/main/java/org/jboss/da/rest/reports/Reports.java [hunk 13]
```diff
@@ -240,9 +278,22 @@ public Response builtReport(BuiltReportRequest request) {
...
-            return Response.status(Status.INTERNAL_SERVER_ERROR).entity(e).build();
+            return Response.status(Status.INTERNAL_SERVER_ERROR)
+                    .entity(new ErrorMessage(ErrorMessage.eType.POM_ANALYSIS, e.getMessage()))
+                    .build();
+        } catch (CommunicationException e) {
+            log.error("Exception thrown in scm endpoint", e);
```
Comment (janinko) [inline reports-rest/src/main/java/org/jboss/da/rest/reports/Reports.java]: This is  exceptiont in communication, not SCM

**Comment by janinko** [comment]
:+1: looks good. You can squash and merge

**Comment by janinko** [comment]
retest this please

**Comment by janinko** [comment]
retest this please

File: reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java (hunk 2 above)
Comment (jomrazek) [inline reports-rest/src/main/java/org/jboss/da/rest/listings/Artifacts.java]: There are two messages for internal server error "There was some internal error"  and "Unexpected server error occurred" I think second sounds better, can you use it on other places too? @jbrazdil?

**Comment by jomrazek** [comment]
:+1:

**Comment by janinko** [comment]
retest this please

**Comment by janinko** [comment]
Wrong formating.

**Pull request merged**